README.md
# External loader for STM32G071 and W25Q64
I needed a file system for a small web server managed by an STM32G071 microcontroller. This file system comes in the form of a single file to be copied into an SPI flash managed by the MCU.
How to copy this file in the flash?

## What is an external loader ?
STM has a solution with _STM32CubeProgrammer_, which allows applications to be flashed in the MCU flash, but also to use external loaders.
An external loader (EL) is a small specific application intended to be executed in the RAM of the MCU (therefore without touching its flash application) and which allows operations other than flashing the MCU. The API is very simple:
- Initialization
- Erasing a sector
- Global erase
- Writing of sectors

The EL is written to run on a given microcontroller to access a particular flash on a particular SPI port. It is therefore necessary to develop the EL adapted to each particular application.

## Achievement
My need: Manage a Winbond W25Q64 flash in simple SPI.

There are EL examples provided by STM (https://github.com/STMicroelectronics/stm32-external-loader), but they concern the brand's development boards and therefore QPSPI flashes (quick access with the flash mapped into the memory space of the MCU). Moreover, these examples are reduced to the simplest expression: no way to run them outside of _STM32CubeProgrammer_, and therefore to debug.
See the MOOC: https://www.st.com/content/st_com/en/support/learning/stm32-education/stm32-moocs/external_QSPI_loader.html

After searching everywhere I wrote this external loader.
The application can be generated in two modes:
- In standard mode (with an _.elf_ extension) to be placed in the MCU flash. This allows you to debug comfortably and verify that the 4 functions of the EL are working, as well as the initialization of the MCU and the SPI. Once everything is working, you can switch to EL mode.
- In EL mode (with a _.stldr_ extension), therefore intended to be placed in the MCU's RAM. And in this case you should know that the startup code will not be executed, and therefore the _main()_ either. All initialization must be done in the _Init()_ function.

The change of mode is done by changing the value of _WITH_MAIN_ in _global.h_, and by **changing the script of the linker** (otherwise the behavior can be very strange!). See comments in _global.h_.

A UART is used (without interrupt, that's important) to trace EL API function calls. This UART is connected to the VCP port of the ST-LINK probe, and the traces are displayed in a PuTTY terminal.
Incidentally, an LED indicates the status of the application in standard mode.

The code uses LL rather than HAL, to generate smaller code. It is important to optimize the RAM occupation, which must contain the application and the buffers used by STM32CubeProgrammer to transmit the sectors to be written. And in some MCUs the RAM is small…

This is a TrueSTUDIO project, because I like using it, but the project can be imported into STM32CubeIDE without issue.

Apparently it is important to respect the EL linker script provided by the MOOC. During the development, before finding this MOOC, I had strange behaviors.

Once the EL has been generated, it must be placed where _STM32CubeProgrammer_ can find it:

_C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin\ExternalLoader_

For use with STM32CubeProgrammer refer to its documentation.


This EL and [MFL file system](https://adastra-soft.com/a-minimalist-file-system/) are among my basic tools for any small WEB server application on MCU.


project/changelog.txt

For the last release of AdAstra-RTK visit AdAstra-Soft . com

AASun_External_Loader is an external loader for STM32CubeProgrammer
It is configured for development and debug.
Can be compiled as standard version (with G071xx.ld) for test and debug
or as external loader with extLoader.ld

The project is configured for a G071CB custom board

--------------------------------------------------------
2023/01		Project external loader W25Q64_AASUN for AASun

23/05/24	1st release

--------------------------------------------------------


project/Application/global.h
/*
----------------------------------------------------------------------

	Alain Chebrou

	global.h	Definitions to include in every source files

	When		Who	What
	21/05/23	ac	Creation

----------------------------------------------------------------------
*/

#if ! defined GLOBAL_H_
#define GLOBAL_H_
//--------------------------------------------------------------------------------

#include <stm32g0xx.h>
#include <stdint.h>

#define	VERSION			" V1.0"


// BEWARE: When you change WITH_MAIN value
//         don't forget to set the according linker script (in project linker general properties)
//         G071xx.ld    for test
//         extLoader.ld for external loader

// Value 1 to compile for test, 0 to compile for external loader
#define	WITH_MAIN		0



#define __ALWAYS_STATIC_INLINE	__attribute__ ((always_inline)) static inline
#define	BSP_ATTR_USED			__attribute__ ((used))

void	bspAssertFailed 	(uint8_t * pFileName, uint32_t line) ;
#define AA_ASSERT(expr) 	((expr) ? (void)0 : bspAssertFailed ((uint8_t *)__FILE__, __LINE__))

// .bss section limits
extern	uint32_t			_sbss ;
extern	uint32_t			_ebss ;

void	initTick			(void) ;
void	aaTaskDelay			(uint32_t ms) ;

void	bspSystemClockInit	(void) ;

void 	uartInit			(void) ;
void	uartPutChar			(char cc) ;
void	uartPuts			(char * pStr) ;
void	uartPrintI			(int32_t value) ;
void	uartPrintX			(uint32_t value, uint32_t width) ;

typedef enum
{
	AA_ENONE				= 0u,	// No error
	AA_EFAIL,						// Generic error
	AA_EARG,						// Bad argument value
	AA_ETIMEOUT,					// Timeout
	AA_EDEPLETED,					// Depleted resource
	AA_ESTATE,						// Operation not allowed for this object state
	AA_EWOULDBLOCK,					// Non blocking function would block
	AA_EFLUSH,						// Object flushed
	AA_ENOTALLOWED,					// Not allowed (e.g. from ISR)
	AA_EMEMORY						// Memory allocation error

} aaError_t ;

//--------------------------------------------------------------------------------
#endif	// GLOBAL_H_


project/Application/Dev_Inf.c
/*
 * Dev_Inf.c
 *
 */
#include <stdint.h>
#include "Dev_Inf.h"
#include "w25q.h"

/* This structure contains information used by ST-LINK Utility to program and erase the device */
#if defined (__ICCARM__)
__root struct StorageInfo const StorageInfo  =
#else
struct StorageInfo __attribute__((section(".Dev_info"))) const StorageInfo =
#endif
{
	"W25Q64_AASun", 	 	 			// Device Name + Board Name
	SPI_FLASH,                  		// Device Type
	W25Q_FLASH_ADDRESS,                	// Device Start Address
	W25Q_FLASH_SIZE,                 	// Device Size in Bytes
	W25Q_PAGE_SIZE,                     // Programming Page Size
	0xFF,                               // Initial Content of Erased Memory

	// Specify Size and Address of Sectors
	{
		{ (W25Q_FLASH_SIZE / W25Q_SECTOR_SIZE),  // Sector Numbers,
				 W25Q_SECTOR_SIZE },             // Sector Size

		{ 0x00000000, 0x00000000 }               // End of list
	}
};



project/Application/Dev_Inf.h

#ifndef DEV_INF_H_
#define DEV_INF_H_

#define 	MCU_FLASH 	1
#define 	NAND_FLASH  2
#define 	NOR_FLASH   3
#define 	SRAM        4
#define 	PSRAM       5
#define 	PC_CARD     6
#define 	SPI_FLASH   7
#define 	I2C_FLASH   8
#define 	SDRAM       9
#define 	I2C_EEPROM  10

#define SECTOR_NUM 2				// Max Number of Sector types

struct DeviceSectors
{
  unsigned long		SectorNum;     // Number of Sectors
  unsigned long		SectorSize;    // Sector Size in Bytes
};

struct StorageInfo
{
   char       	  DeviceName[100];			// Device Name and Description
   unsigned short DeviceType;				// Device Type: ONCHIP, EXT8BIT, EXT16BIT, ...
   unsigned long  DeviceStartAddress;		// Default Device Start Address
   unsigned long  DeviceSize;				// Total Size of Device
   unsigned long  PageSize;					// Programming Page Size
   unsigned char  EraseValue;				// Content of Erased Memory
   struct 	  DeviceSectors	 sectors[SECTOR_NUM];
};

extern struct StorageInfo const StorageInfo ;

#define LOADER_OK	0x1
#define LOADER_FAIL	0x0

int		Init		(void) ;
int		Write		(uint32_t Address, uint32_t Size, uint8_t* buffer) ;
int		SectorErase	(uint32_t EraseStartAddress, uint32_t EraseEndAddress) ;
int		MassErase	(void) ;

#endif /* DEV_INF_H_ */


project/Application/Loader_Src.c

#include	"global.h"
#include	"w25q.h"
#include	"Dev_inf.h"

#include	<string.h>					// memset

/**
 * @brief  System initialization.
 * @param  None
 * @retval  LOADER_OK = 1	: Operation succeeded
 * @retval  LOADER_FAIL = 0	: Operation failed
 */

// The Init function is mentioned by Entry Point in the linker script
// and directly called by STM32CubeProgrammer, so the startup code is not executed

int Init (void)
{

	*(uint32_t*) 0xE000EDF0 = 0xA05F0000 ; // enable interrupts in debug

    // Init .bss section to Zero
    memset (& _sbss, 0, & _ebss - & _sbss);

	SystemInit () ;

#if (WITH_MAIN == 0)
/* ADAPTATION TO THE DEVICE
 *
 * change VTOR setting for H7 device
 * SCB->VTOR = 0x24000000 | 0x200;
 *
 * change VTOR setting for other devices
 * SCB->VTOR = 0x20000000 | 0x200;
 */
	// Only if compiled as external loader
	SCB->VTOR = 0x20000000 | 0x200 ;
#endif

	bspSystemClockInit () ;
	initTick  () ;
	W25Q_Init () ;
	uartInit  () ;

	uartPuts    ("I") ;
	uartPrintX  (StorageInfo.DeviceStartAddress, 9) ;
	uartPrintX  (StorageInfo.DeviceSize, 9) ;
	uartPuts    (VERSION) ;
	uartPutChar ('\n') ;

	return LOADER_OK ;
}

/**
 * @brief   Program memory.
 * @param   Address: page address
 * @param   Size   : size of data
 * @param   buffer : pointer to data buffer
 * @retval  LOADER_OK = 1	: Operation succeeded
 * @retval  LOADER_FAIL = 0	: Operation failed
 */
int Write(uint32_t Address, uint32_t Size, uint8_t* buffer)
{
	// Page (266 B) write time: 0.7 - 3 ms

	uartPutChar ('W') ;	uartPrintX  (Address, 9) ; uartPrintX  (Size, 9) ; uartPutChar ('\n') ;

	W25Q_Write (buffer, Address, Size) ;
	return LOADER_OK;
}

int Read(uint32_t Address, uint32_t Size, uint8_t *Buffer)
{
	uartPuts ("R") ;	uartPrintX  (Address, 9) ; uartPrintX  (Size, 9) ; uartPutChar ('\n') ;

	W25Q_Read (Buffer, Address, Size) ;
	return LOADER_OK;
}

/**
 * @brief   Sector erase.
 * @param   EraseStartAddress :  erase start address
 * @param   EraseEndAddress   :  erase end address
 * @retval  LOADER_OK = 1		: Operation succeeded
 * @retval  LOADER_FAIL = 0	: Operation failed
 */
int SectorErase (uint32_t EraseStartAddress, uint32_t EraseEndAddress)
{
	// 4K sector erase time: 60 - 400 ms
	uartPuts ("SE") ;	uartPrintX  (EraseStartAddress, 9) ; uartPrintX  (EraseEndAddress, 9) ; uartPutChar ('\n') ;

	while (EraseStartAddress <= EraseEndAddress)
	{
		uartPutChar ('E') ; uartPrintX  (EraseStartAddress, 9) ; uartPutChar ('\n') ;
		W25Q_EraseSector (EraseStartAddress) ;
		EraseStartAddress += W25Q_SECTOR_SIZE ;
	}
	return LOADER_OK;
}

/**
 * Description :
 * Mass erase of external flash area
 * Optional command - delete in case usage of mass erase is not planed
 * Inputs    :
 *      none
 * outputs   :
 *     none
 * Note: Optional for all types of device
 */
int MassErase (void)
{
	// Do not erase all the Flash: the 1st MB is reserved
	// Erase using 64K blocs
	// 64K block erase time: 150 - 2000 ms

	uint32_t	EraseStartAddress = W25Q_FLASH_ADDRESS ;
	uint32_t	EraseEndAddress   = EraseStartAddress + W25Q_FLASH_SIZE ;

	uartPuts ("Mass") ; uartPrintX  (EraseStartAddress, 9) ; uartPrintX  (EraseEndAddress, 9) ; uartPutChar ('\n') ;

	while (EraseStartAddress < EraseEndAddress)
	{
		W25Q_EraseBlock64 (EraseStartAddress) ;
		EraseStartAddress += W25Q_BLOCK64_SIZE ;
		uartPutChar ('M') ; uartPrintX  (EraseStartAddress, 9) ; uartPutChar ('\n') ;
	}
	return LOADER_OK;
}

#if (0)
/**
 * Description :
 * Calculates checksum value of the memory zone
 * Inputs    :
 *      StartAddress  : Flash start address
 *      Size          : Size (in WORD)
 *      InitVal       : Initial CRC value
 * outputs   :
 *     R0             : Checksum value
 * Note: Optional for all types of device
 */
uint32_t CheckSum (uint32_t StartAddress, uint32_t Size, uint32_t InitVal)
{
	uint8_t missalignementAddress = StartAddress % 4;
	uint8_t missalignementSize = Size;
	uint32_t cnt;
	uint32_t Val;

	StartAddress -= StartAddress % 4;
	Size += (Size % 4 == 0) ? 0 : 4 - (Size % 4);

	for (cnt = 0; cnt < Size; cnt += 4)
	{
		Val = *(uint32_t*) StartAddress;
		if (missalignementAddress)
		{
			switch (missalignementAddress)
			{
			case 1:
				InitVal += (uint8_t) (Val >> 8 & 0xff);
				InitVal += (uint8_t) (Val >> 16 & 0xff);
				InitVal += (uint8_t) (Val >> 24 & 0xff);
				missalignementAddress -= 1;
				break;
			case 2:
				InitVal += (uint8_t) (Val >> 16 & 0xff);
				InitVal += (uint8_t) (Val >> 24 & 0xff);
				missalignementAddress -= 2;
				break;
			case 3:
				InitVal += (uint8_t) (Val >> 24 & 0xff);
				missalignementAddress -= 3;
				break;
			default:
				break ;
			}
		}
		else if ((Size - missalignementSize) % 4 && (Size - cnt) <= 4)
		{
			switch (Size - missalignementSize)
			{
			case 1:
				InitVal += (uint8_t) Val;
				InitVal += (uint8_t) (Val >> 8 & 0xff);
				InitVal += (uint8_t) (Val >> 16 & 0xff);
				missalignementSize -= 1;
				break;
			case 2:
				InitVal += (uint8_t) Val;
				InitVal += (uint8_t) (Val >> 8 & 0xff);
				missalignementSize -= 2;
				break;
			case 3:
				InitVal += (uint8_t) Val;
				missalignementSize -= 3;
				break;
			default:
				break ;
			}
		}
		else
		{
			InitVal += (uint8_t) Val;
			InitVal += (uint8_t) (Val >> 8 & 0xff);
			InitVal += (uint8_t) (Val >> 16 & 0xff);
			InitVal += (uint8_t) (Val >> 24 & 0xff);
		}
		StartAddress += 4;
	}

	return (InitVal);
}

/**
 * Description :
 * Verify flash memory with RAM buffer and calculates checksum value of
 * the programmed memory
 * Inputs    :
 *      FlashAddr     : Flash address
 *      RAMBufferAddr : RAM buffer address
 *      Size          : Size (in WORD)
 *      InitVal       : Initial CRC value
 * outputs   :
 *     R0             : Operation failed (address of failure)
 *     R1             : Checksum value
 * Note: Optional for all types of device
 */
BSP_ATTR_USED uint64_t Verify (uint32_t MemoryAddr, uint32_t RAMBufferAddr, uint32_t Size, uint32_t missalignement)
{
	uint32_t VerifiedData = 0, InitVal = 0;
	uint64_t checksum;
	Size *= 4;

	checksum = CheckSum((uint32_t) MemoryAddr + (missalignement & 0xf), Size - ((missalignement >> 16) & 0xF), InitVal);
	while (Size > VerifiedData)
	{
		if (*(uint8_t*) MemoryAddr++  !=  *((uint8_t*) RAMBufferAddr + VerifiedData))
		{
			return ((checksum << 32) + (MemoryAddr + VerifiedData));
		}
		VerifiedData++;
	}

	return (checksum << 32);
}
#endif


project/Application/SystemClock_Config.c
/*
----------------------------------------------------------------------

	AdAstra - Real Time Kernel

	Alain Chebrou

	SystemClock_Config.c		The function to initialize MCU clocks
								Example for:
									STM32G071
								 These functions are examples and should be changed as needed
	When		Who	What
	18/01/21	ac	Created

----------------------------------------------------------------------
*/

#include	"stm32g0xx_ll_rcc.h"
#include	"stm32g0xx_ll_system.h"
#include	"stm32g0xx_ll_pwr.h"
#include	"stm32g0xx_ll_bus.h"
#include	"stm32g0xx_ll_utils.h"

// ----------------------------------------------------------------------------
//	SystemClock_Config() code was generated by STM32CubeMX
//	This function must be adapted: configure the clock in accordance with your needs

//	The function must:
//		Not initialize SysTick
//		Initialize peripheral clocks
//		Set TIMPRE if requested


//	Use 8 MHz HSE as system clock source, system clock is 64 MHz
//	Don't forget to set HSE_VALUE=8000000 in the tool settings

void SystemClock_Config (void)
{
	LL_FLASH_SetLatency(LL_FLASH_LATENCY_2);
	while(LL_FLASH_GetLatency() != LL_FLASH_LATENCY_2)
	{
	}

	// HSE configuration and activation
	LL_RCC_HSE_Enable();
	while(LL_RCC_HSE_IsReady() != 1)
	{
	}

	// Main PLL configuration and activation
	LL_RCC_PLL_ConfigDomain_SYS(LL_RCC_PLLSOURCE_HSE, LL_RCC_PLLM_DIV_1, 16, LL_RCC_PLLR_DIV_2);
	LL_RCC_PLL_Enable();
	LL_RCC_PLL_EnableDomain_SYS();
	while(LL_RCC_PLL_IsReady() != 1)
	{
	}

	// Set AHB prescaler
	LL_RCC_SetAHBPrescaler(LL_RCC_SYSCLK_DIV_1);

	// Sysclk activation on the main PLL
	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_PLL);
	while(LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_PLL)
	{
	}

	// Set APB1 prescaler
	LL_RCC_SetAPB1Prescaler(LL_RCC_APB1_DIV_1);

	LL_RCC_SetUSARTClockSource(LL_RCC_USART1_CLKSOURCE_SYSCLK);
#if (defined RCC_CCIPR_USART2SEL)
	LL_RCC_SetUSARTClockSource(LL_RCC_USART2_CLKSOURCE_SYSCLK);
#endif
	LL_RCC_SetADCClockSource(LL_RCC_ADC_CLKSOURCE_SYSCLK);
}

// ----------------------------------------------------------------------------
//	The purpose of this function is to provide a prologue and an epilogue to SystemClock_Config().

void	bspSystemClockInit ()
{
	// These initialization may be required if SystemClock_Config() configure LSE, or other
	LL_APB2_GRP1_EnableClock (LL_APB2_GRP1_PERIPH_SYSCFG);
	LL_APB1_GRP1_EnableClock (LL_APB1_GRP1_PERIPH_PWR);

	SystemClock_Config () ;

	// Stop SysTick if it is started by SystemClock_Config()
	SysTick->CTRL = 0 ;

	SystemCoreClockUpdate () ;
}

// ----------------------------------------------------------------------------


project/Application/main.c
/*
----------------------------------------------------------------------

	AASun - External loader for web pages

	Alain Chebrou

	main.c		Entry point for tests

	When		Who	What
	21/05/23	ac	Creation

----------------------------------------------------------------------
*/

#include	"global.h"
#include	"gpiobasic.h"
#include	"w25q.h"
#include	"Dev_inf.h"
#include	"stm32g0xx_ll_usart.h"
#include	"stm32g0xx_ll_rcc.h"
#include	"stm32g0xx_ll_bus.h"
#include	"stm32g0xx_ll_utils.h"

#include	"string.h"	// For memset, memcmp

// The GPIO descriptor for the LED
const gpioPinDesc_t	ledDesc = { 'B',	12,	0, AA_GPIO_MODE_OUTPUT_PP | AA_GPIO_SPEED_LOW } ;

//--------------------------------------------------------------------------------
//--------------------------------------------------------------------------------
// aaTaskDelay() is used by w25q.c, assert_failed(), main()

void	initTick (void)
{
	LL_InitTick (__LL_RCC_CALC_HCLK_FREQ (SystemCoreClock, LL_RCC_GetAHBPrescaler()), 1000) ;	// For delays
}

void aaTaskDelay (uint32_t ms)
{
	LL_mDelay (ms) ;
}

//--------------------------------------------------------------------------------
//--------------------------------------------------------------------------------
// main() is not used when compiled as external loader: WITH_MAIN 0
// main() is only used when compiled for test:          WITH_MAIN 1

// BEWARE: When you change WITH_MAIN value
//         don't forget to set the according linker script (in project linker properties)

#if (WITH_MAIN == 0)

	int main (void)
	{
		while (1)
		{
		}
	}

#else

static	uint8_t		buffer_test   [W25Q_SECTOR_SIZE] ;
static	uint8_t		sectorBuffer  [W25Q_SECTOR_SIZE] ;

// When compiled as standard application, main() allows to exercise
// all the API functions of the external loader

int main (void)
{
	uint32_t	var ;
	uint32_t	sectorAddress ;

	__disable_irq () ;
	Init () ;

	// This is a test of the Loader_Src functions and the Flash driver

	// Initialize the flash sector data
	for (var = 0 ; var < W25Q_SECTOR_SIZE ; var++)
	{
		buffer_test [var] = var & 0xff ;
	}
/*
	// SectorErase() test, erase all sectors
	uartPuts    ("Sector erase ") ;
	uartPrintX  (StorageInfo.DeviceStartAddress, 9) ;
	uartPrintX  (StorageInfo.DeviceStartAddress+StorageInfo.DeviceSize-1, 9) ;
	uartPutChar ('\n') ;

	SectorErase (StorageInfo.DeviceStartAddress, StorageInfo.DeviceStartAddress+StorageInfo.DeviceSize-1) ;
	uartPuts ("Erase OK\n") ;
*/
	// Mass erase test
	MassErase () ;
	uartPuts  ("\nOK\n") ;

	// Write sectors test
	sectorAddress = StorageInfo.DeviceStartAddress ;
	for (var = 0 ; var < W25Q_FLASH_SIZE / W25Q_SECTOR_SIZE ; var++)
	{
		if (Write (sectorAddress, W25Q_SECTOR_SIZE, buffer_test) != LOADER_OK)
		{
			uartPuts ("Write error\n") ;
			AA_ASSERT (0) ; // Error detected, LED blink fast
		}
		sectorAddress += W25Q_SECTOR_SIZE ;
		gpioPinToggle (& ledDesc) ;
	}
	uartPutChar ('\n') ;

	// Verify Flash data
	sectorAddress = StorageInfo.DeviceStartAddress ;
  	for (var = 0 ; var < W25Q_FLASH_SIZE / W25Q_SECTOR_SIZE ; var++)
  	{
  		W25Q_Read (sectorBuffer, sectorAddress, W25Q_SECTOR_SIZE) ;
  		if (memcmp (buffer_test, sectorBuffer, W25Q_SECTOR_SIZE) != 0)
  		{
			uartPuts ("Verify  error\n") ;
  			AA_ASSERT (0) ;		// Error detected, LED blink fast
  		}
  		gpioPinToggle (& ledDesc) ;
  		uartPutChar ('-') ;
	}
  	uartPutChar ('\n') ;
	uartPuts ("Test OK\n") ;
  	// End of test


  	// If test OK then the LED blinks slowly
	while (1)
	{
		aaTaskDelay (500) ;
		gpioPinToggle (& ledDesc) ;
	}
}
#endif

//--------------------------------------------------------------------------------
//--------------------------------------------------------------------------------
//	Simple polling driver for USART2

#define	BAUD_RATE	57600

static const gpioPinDesc_t	uartTxDesc = { 'A', 2, 1, AA_GPIO_MODE_ALTERNATE | AA_GPIO_PUSH_PULL | AA_GPIO_SPEED_LOW } ;
static const gpioPinDesc_t	uartRxDesc = { 'A', 3, 1, AA_GPIO_MODE_ALTERNATE | AA_GPIO_PUSH_PULL | AA_GPIO_SPEED_LOW } ;

//--------------------------------------------------------------------------------
// The LL_USART_Init function is huge, 600 bytes, but easy to use...

void uartInit (void)
{
	LL_USART_InitTypeDef USART_InitStruct = { 0 } ;

	LL_APB1_GRP1_EnableClock  (LL_APB1_GRP1_PERIPH_USART2) ;
	LL_APB1_GRP1_ForceReset   (LL_APB1_GRP1_PERIPH_USART2) ;
	LL_APB1_GRP1_ReleaseReset (LL_APB1_GRP1_PERIPH_USART2) ;

	gpioConfigurePin (& uartTxDesc) ;
	gpioConfigurePin (& uartRxDesc) ;

	USART_InitStruct.PrescalerValue			= LL_USART_PRESCALER_DIV1 ;
	USART_InitStruct.BaudRate				= BAUD_RATE ;
	USART_InitStruct.DataWidth				= LL_USART_DATAWIDTH_8B ;
	USART_InitStruct.StopBits				= LL_USART_STOPBITS_1 ;
	USART_InitStruct.Parity					= LL_USART_PARITY_NONE ;
	USART_InitStruct.TransferDirection		= LL_USART_DIRECTION_TX_RX ;
	USART_InitStruct.HardwareFlowControl	= LL_USART_HWCONTROL_NONE ;
	USART_InitStruct.OverSampling			= LL_USART_OVERSAMPLING_16 ;
	LL_USART_Init (USART2, & USART_InitStruct) ;

	USART2->CR1 |= USART_CR1_UE ;			// Enable USART
	// Wait for TEAK and REAK flags set
	while ((USART2->ISR & (USART_ISR_TEACK | USART_ISR_REACK)) != (USART_ISR_TEACK | USART_ISR_REACK))
	{
	}
}

//--------------------------------------------------------------------------------
// Send 1 char to UART

void uartPutChar (char cc)
{
	volatile USART_TypeDef	* uartBase = USART2 ;	// UARTx
	uint32_t				flagLF = 0u ;

	// If cc is LF, then send CR-LF
	if (cc == '\n')
	{
		flagLF = 1u ;
		cc = '\r' ;
	}

	uartBase->TDR = (uint16_t) cc ;
	// Wait until the FIFO is not full
	while ((uartBase->ISR & USART_ISR_TXE_TXFNF) == 0u)
	{
	}

	if (flagLF != 0u)
	{
		uartBase->TDR = '\n' ;
		while ((uartBase->ISR & USART_ISR_TXE_TXFNF) == 0u)
		{
		}
	}
}

//--------------------------------------------------------------------------------
//	Send 0 terminated string

void uartPuts (char * pStr)
{
	while (* pStr != 0)
	{
		uartPutChar (* pStr++) ;
	}
}

//--------------------------------------------------------------------------------
// Simple decimal number print without format

void	uartPrintI (int32_t value)
{
	char		str [12] ;
	char		* ptr ;
	char		chr ;
	uint32_t	neg = 0 ;

	if (value < 0)
	{
		neg = 1 ;
		value = - value ;
	}

	ptr = & str [sizeof (str) - 1] ;
	* ptr = 0 ;
	do
	{
		chr = (char) ((value % 10) + '0') ;
		*--ptr = chr ;
		value /= 10 ;
	} while (value != 0) ;

	if (neg != 0)
	{
		uartPutChar ('-') ;
	}

	while (* ptr != 0)
	{
		uartPutChar (* ptr++) ;
	}
}

//--------------------------------------------------------------------------------
// Simple hex number print with user defined width

void	uartPrintX (uint32_t value, uint32_t width)
{
	char		str [12] ;
	char		* ptr ;
	char		chr ;

	ptr = & str [sizeof (str) - 1] ;
	* ptr = 0 ;
	do
	{
		chr = (char) ((value & 0x0F) + '0') ;
		if (chr > '9')
		{
			chr += 7 ;
		}
		*--ptr = chr ;
		value >>= 4 ;
		if (width > 0)
		{
			width--;
		}
	} while (value != 0) ;

	while (width > 0)
	{
		*--ptr = ' ' ;
		width-- ;
	}

	while (* ptr != 0)
	{
		uartPutChar (* ptr++) ;
	}
}

//--------------------------------------------------------------------------------
//--------------------------------------------------------------------------------
//	LL Support functions

//--------------------------------------------------------------------------------
//	AA_ASSERT support function

void	bspAssertFailed (uint8_t * pFileName, uint32_t line)
{
	(void) pFileName ;
	(void) line ;

	// Check if debug is enabled.
	// If debug is enabled, use bkpt instruction to jump to debugger
	// If debug is not enabled, a bkpt instruction triggers a HardFault exception
	#if (__CORTEX_M != 0)
	{
		// No CoreDebug on M0 / M0+
		if ((CoreDebug->DHCSR & CoreDebug_DHCSR_C_DEBUGEN_Msk) != 0u)
		{
			// Debug enabled, can use BKPT instruction
			BSP_DEBUG_BKPT () ;
		}
	}
	#endif

	// No debugger: loop with fast LED blink
	__disable_irq () ;
	while (1)
	{
		aaTaskDelay (200) ;
		gpioPinToggle (& ledDesc) ;
	}
}

//--------------------------------------------------------------------------------
// Support for STMicroelectronics stdperiph, HAL or LL code (stm32_assert.h)
// Not the same assert_failed() prototype in all families...
// For F4 / F1 / G0

void assert_failed (uint8_t * file, uint32_t line)
{
	bspAssertFailed (file, line) ;
}

//--------------------------------------------------------------------------------
